core/src/main/scala/org/apache/spark/scheduler/bus/ListenerBusQueueImpl.scala

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.scheduler.bus
+
+import com.codahale.metrics.Timer
+import scala.reflect.ClassTag
+
+import org.apache.spark.scheduler.{SparkListenerEvent, SparkListenerEventDispatcher, SparkListenerInterface}
+import org.apache.spark.scheduler.bus.ListenerBusQueue.{FixGroupOfListener, ModifiableGroupOfListener}
+
+// For generic message processor (like event logging)
+private[scheduler] class ProcessorListenerBusQueue(
```

Comment:
I'm still confused about why you need 2 implementations. Why doesn't `ListenerBusQueue` work for everybody? And why shouldn't it?

I need more time to actually grok all this code, but like Wenchen suggested before, this is a big change and it would benefit from a more detailed explanation of exactly how you're organizing the hierarchy of listener, groups, etc. Your PR description only explains which queues you created, but not any of the changes that were needed to achieve that.

If it makes it easier, you can create a `README.md` file with a longer explanation for how things are organized. (for example, check `ommon/network-common/src/main/java/org/apache/spark/network/crypto` where I added a README to explain details of what that whole body of code is doing).